### DISCUSSION TITLE
ObjectCache: how to handle child objects without their own id?

### DISCUSSION THREAD

**Discussion by ludo** [rmosolgo/graphql-ruby#5110] [Q&A] [answered]
We're setting up ObjectCache in our GraphQL API. One thing I'm running into is that we have a few GraphQL types that are virtual (not backed by an identifiable object) and I'm not sure how to deal with them. If I do nothing GraphQL Ruby will be unhappy because `id_from_object` will cause an exception.

It seems like [Caching lists and connections](https://graphql-ruby.org/object_cache/caching#caching-lists-and-connections) and [Object Dependencies](https://graphql-ruby.org/object_cache/caching#object-dependencies) are relevant to this problem but I can't wrap my head around how to apply them. See code below, I think I'm looking for a way to call the `parent` object in `FormFieldType`.

Here's an excerpt of the code I'm working with:

```ruby
module Types
  class FormType < Types::BaseObject
    field :fields, [Types::FormFieldType], null: false
  end
  
  class FormFieldType < Types::BaseObject
    field :name, String, null: false
  end
end
```

`FormType` resolves to a `Form` ActiveRecord model. This model has a `fields` attribute which is an array of hashes (JSON in the database). For example: `#<Form id: 1, fields: [{"name"=>"FOO"}, {"name"=>"BAR"}]>`.

At this point when we query a form with its fields it will raise an error because fields aren't 'real' objects with ids:

```graphql
query {
  forms {
    nodes {
      fields {
        name
      }
    }
  }
}
```

```
Failure/Error: Base64.strict_encode64(object.to_global_id.to_s)

NoMethodError:
  undefined method `to_global_id' for #<...>
```

I kind of just want to tell FormFieldType to be cached as part of its parent if that makes sense?

**Comment by rmosolgo** [comment, marked as answer]
Hey, thanks for the great question. 

Yes, I think `cache_dependencies_for` is the way to go, you could implement it like this: 

```ruby 
class Types::FormFieldType < Types::BaseObject
  def self.cache_dependencies_for(form_field, context)
    form_field.parent 
  end  
  
  # ... 
end 
```


Can you return the form field's `form` object easily? The most fail-safe would be to use an application-defined method on the form field (or add one if necessary, add `"form" => form` to the hash, etc). That way, you could be sure that the form field would have access to the right parent, no matter what. 

In a pinch, you could use GraphQL-Ruby's "scoped context" field to accomplish this, but I don't recommend it because scoped context breaks if the query doesn't meet the assumptions of the Ruby code. (That is, if the incoming query was a different structure than assumed, weird things can happen since the query tree was different.) In any case, here it is: https://graphql-ruby.org/queries/executing_queries.html#scoped-context

Does that help?

**Reply by ludo** [reply]
> Can you return the form field's form object easily? The most fail-safe would be to use an application-defined method on the form field (or add one if necessary, add "form" => form to the hash, etc). That way, you could be sure that the form field would have access to the right parent, no matter what.

Ah, thanks, that's a good idea. I was initially trying to access the parent through `context` (I don't think that's available?). Your suggestion seems fine though. This works and is simple enough:

```ruby
module Types
  class FormType < Types::BaseObject
    field :fields, [Types::FormFieldType], null: false
    def fields
      object.fields.map { |field| field.merge({ parent: object }) }
    end
  end
  
  class FormFieldType < Types::BaseObject
    def self.cache_dependencies_for(form_field)
      form_field[:parent]
    end

    field :name, String, null: false
  end
end
```

Thanks for your help!
